File: leetcode/total-hamming-distance.py
#!/usr/bin/env python
# coding:utf-8
# Copyright (C) dirlt

from typing import List


class Solution:
    def totalHammingDistance(self, nums: List[int]) -> int:
        bits = [0] * 32
        for x in nums:
            for i in range(32):
                if x & (1 << i):
                    bits[i] += 1
        n = len(nums)
        ans = 0
        for x in bits:
            ans += x * (n - x)
        return ans

File: leetcode/cpp/total-hamming-distance.cpp
/* coding:utf-8
 * Copyright (C) dirlt
 */

#include <vector>
using namespace std;

class Solution {
   public:
    int totalHammingDistance(vector<int>& nums) {
        int bits[32] = {0};
        for (int k = 0; k < nums.size(); k++) {
            int n = nums[k];
            for (int p = 0; p < 32; p++) {
                if (n & (1 << p)) {
                    bits[p] += 1;
                }
            }
        }
        int count = 0;
        for (int p = 0; p < 32; p++) {
            count += (nums.size() - bits[p]) * bits[p];
        }
        return count;
    }
};
